Create guide for modeling go libraries (#233)
github/codeql-go: docs/language/learn-ql/go/library-modeling-go.rst

> Modeling Go libraries for CodeQL

Reviewer: Two general points about the title: 

- We don't really need "for CodeQL" in the title (this should be clear once the file is put into the right context) 
- Looking at the content, it seems like this guide is specifically about data flow. Is that a correct interpretation? If so, it might be more descriptive to change the title to something like "Analyzing data flow in Go" (in line with the other languages, e.g. https://help.semmle.com/QL/learn-ql/javascript/dataflow.html). If that doesn't fit, ignore me 🙂

Reviewer 2: This guide is a little more focused than our usual "Analyzing data flow in \<language>" articles, so that title probably wouldn't be ideal. However, I agree that putting something about data flow into the title would be good.

Author: Thanks for pointing out that page, which I wasn't aware of. It's not exactly the same topic. We always just refer to modelling a library, but I guess it's a model of the data flow (or technically the taint flow, but they're closely related). I've made the title "Modeling data flow in Go libraries" now.

================================

> When analyzing a Go program, CodeQL does not examine the source code for
> external packages. To track the flow of untrusted data through them you
> need to create a model of the library. Existing models can be found in
> ``ql/src/semmle/go/frameworks/``, and are a good source of examples. You
> should make a new file in that folder, named after the library.

Reviewer: ```suggestion
When analyzing a Go program, CodeQL does not examine the source code for
external packages. To track the flow of untrusted data through external packages, you
can create a model of the library. 

You can find existing models in the ``ql/src/semmle/go/frameworks/`` folder of the `CodeQL for Go repository <https://github.com/github/codeql-go/tree/main/ql/src/semmle/go/frameworks>`__.
To add a new model, you should make a new file in that folder, named after the library.
```

I've added this as an explicit suggestion, but feel free to modify it if you prefer! 
- We tend to have a 1-2 sentence introduction to use in search/TOC (so I've separated out "When analyzing ... create a model of the library" to act as the intro.) 
- Added an explicit link to the ``frameworks`` folder. 
- Some minor changes to minimise use of potentially ambiguous pronouns and passive voice. (This is something that will aid translation once the docs are part of the GitHub system 🌍)

Author: Great. I've slightly modified it because I realised I was using "package" and "library" interchangeably, which is confusing.

Author: Also, I didn't know out docs will be translated. That's cool.

Reviewer: I know right! We don't touch the translations ourselves, but GitHub docs get translated into various languages using Crowdin.

Sources
-------

To mark a source of data that is controlled by an untrusted user, we
create a class extending ``UntrustedFlowSource::Range``. Inheritance and
the characteristic predicate of the class should be used to specify
exactly the dataflow node that introduces the data. Here is a short
example from ``Mux.qll``.

.. code-block:: ql

     class RequestVars extends DataFlow::UntrustedFlowSource::Range, DataFlow::CallNode {
       RequestVars() { this.getTarget().hasQualifiedName("github.com/gorilla/mux", "Vars") }
     }

This has the effect that all calls to `the function Vars from the
package mux <http://www.gorillatoolkit.org/pkg/mux#Vars>`__ are
treated as sources of untrusted data.

Flow propagation
----------------

> By default, it is assumed that all functions in libraries do not have
> any data flow. To indicate that a particular function does, you need to
> create a class extending ``TaintTracking::FunctionModel`` (or
> ``DataFlow::FunctionModel`` if the untrusted user data is passed on
> without being modified).

Reviewer: ```suggestion
By default, we assume that functions in libraries do not have
any data flow. To indicate that a particular function does have data flow,
create a class extending ``TaintTracking::FunctionModel`` (or
``DataFlow::FunctionModel`` if the untrusted user data is passed on
without being modified).
```

> Inheritance and the characteristic predicate of the class should specify
> the function and a member predicate is needed with the signature
> ``override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp)``
> (or
> ``override predicate hasDataFlow(FunctionInput inp, FunctionOutput outp)``
> if extending ``DataFlow::FunctionModel``). The body should constrain
> ``inp`` and ``outp``.

Reviewer: ```suggestion
Inheritance and the characteristic predicate of the class should specify
the function. The class must also have a member predicate with the signature
``override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp)``
(or
``override predicate hasDataFlow(FunctionInput inp, FunctionOutput outp)``
if extending ``DataFlow::FunctionModel``). The body should constrain
``inp`` and ``outp``.
```

Suggestion to avoid passive voice

``FunctionInput`` is an abstract representation of the inputs to a
function. The options are:

* the receiver (``inp.isReceiver()``)
* one of the parameters (``inp.isParameter(i)``)
* one of the results (``inp.isResult(i)``, or ``inp.isResult`` if there is only one result)

Note that it may seem strange that the result of a function could be
considered as a function input, but it is needed in some cases. For
instance, the function ``bufio.NewWriter`` returns a writer ``bw`` that
buffers write operations to an underlying writer ``w``. If tainted data
is written to ``bw``, then it makes sense to propagate that taint back
to the underlying writer ``w``, which can be modeled by saying that
``bufio.NewWriter`` propagates taint from its result to its first
argument.

Similarly, ``FunctionOutput`` is an abstract representation of the

> inputs to a function. The options are:

Reviewer: ```suggestion
outputs to a function. The options are:
```
typo?

Author: Gah. I made one of those paragraphs by copying the other one and somehow I could never track down all the "input"s and change them to "output"s.

* the receiver (``outp.isReceiver()``)
* one of the parameters (``outp.isParameter(i)``)
* one of the results (``outp.isResult(i)``, or ``outp.isResult`` if there is only one result)

Here is an example from ``Gin.qll``, slightly modified for brevity.

.. code-block:: ql

   private class ParamsGet extends TaintTracking::FunctionModel, Method {
     ParamsGet() { this.hasQualifiedName("github.com/gin-gonic/gin", "Params", "Get") }

     override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp) {
       inp.isReceiver() and outp.isResult(0)
     }
   }

This has the effect that calls to the ``Get`` method with receiver type
``Params`` from the ``gin-gonic/gin`` package allow taint to flow from
the receiver to the first result. In other words, if ``p`` has type

> ``Params`` and taint can flow to it then after the line

Reviewer: ```suggestion
``Params`` and taint can flow to it, then after the line
```

``x := p.Get("foo")`` taint can also flow to ``x``.

Sanitizers
----------

It is not necessary to indicate that library functions are sanitizers -
because their bodies are not analyzed it is assumed that data does not
flow through them.

Sinks
-----

Data-flow sinks are specified by queries rather than by library models.

> What library models can do is to indicate when functions belong to
> special categories of function, which queries can use when specifying
> sinks. Classes representing these special categories are contained in
> ``ql/src/semmle/go/Concepts.qll``, including ones for logger mechanisms,
> HTTP response writers, HTTP redirects and marshaling and unmarshaling
> functions.

Reviewer: ```suggestion
However, you can use library models to indicate when functions belong to
special categories. Queries can then use these models when specifying
sinks. Classes representing these special categories are contained in
``ql/src/semmle/go/Concepts.qll`` in the `CodeQL for Go repository <https://github.com/github/codeql-go/blob/main/ql/src/semmle/go/Concepts.qll>`__, including classes for logger mechanisms,
HTTP response writers, HTTP redirects, and marshaling and unmarshaling
functions.
```

Author: That's the second time I've been asked to insert an Oxford commas. Is it part of the style we should use?

Reviewer: Yep, it's part of the brand guide: https://brand.github.com/content/grammar#commas

Here is a short example from ``Stdlib.qll``, slightly modified for
brevity.

.. code-block:: ql

   private class PrintfCall extends LoggerCall::Range, DataFlow::CallNode {
     PrintfCall() { this.getTarget().hasQualifiedName("fmt", ["Print", "Printf", "Println"]) }

     override DataFlow::Node getAMessageComponent() { result = this.getAnArgument() }
   }

> This has the effect that any call to ``Print``, ``Printf``, or
> ``Println`` in the package ``fmt`` is recognised as a logger call, and
> in any query which uses logger calls as a sink then passing tainted data
> as an argument to ``Print``, ``Printf``, or ``Println`` will create a
> result for the query.

Reviewer: ```suggestion
This has the effect that any call to ``Print``, ``Printf``, or
``Println`` in the package ``fmt`` is recognized as a logger call.
Any query that uses logger calls as a sink will then identify when tainted data 
has been passed as an argument to ``Print``, ``Printf``, or ``Println``.
```

I found this final sentence quite confusing to read, so I've tried to break it into smaller chunks.
It isn't ideal - feel free to correct it if I've misinterpreted the information!
